File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/FraSedA003Mapper.java
package no.nav.melosys.eessi.service.sed.mapper.fra_sed;

import lombok.extern.slf4j.Slf4j;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA003;
import no.nav.melosys.eessi.service.sed.helpers.LandkodeMapper;

@Slf4j
public abstract class FraSedA003Mapper implements NyttLovvalgSedMapper<MedlemskapA003> {
    @Override
    public String hentLovvalgsland(MedlemskapA003 medlemskap) {
        return LandkodeMapper.mapTilNavLandkode(medlemskap.getVedtak().getLand());
    }

    @Override
    public String hentLovvalgsbestemmelse(MedlemskapA003 medlemskap) {
        return medlemskap.getRelevantartikkelfor8832004eller9872009();
    }

    @Override
    public Boolean sedErEndring(MedlemskapA003 medlemskap) {
        var erEndring = !"ja".equalsIgnoreCase(medlemskap.getVedtak().getEropprinneligvedtak());
        log.info("sedErEndring i A003 er {}, med erendringsvedtak: {} og eropprinneligvedtak: {}", erEndring,
            medlemskap.getVedtak().getErendringsvedtak(), medlemskap.getVedtak().getEropprinneligvedtak());
        return erEndring;
    }

    @Override
    public MedlemskapA003 hentMedlemskap(SED sed) {
        return (MedlemskapA003) sed.getMedlemskap();
    }

    @Override
    public boolean erMidlertidigBestemmelse(MedlemskapA003 medlemskap) {
        return "ja".equals(medlemskap.getIsDeterminationProvisional());
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/FraSedA003Mapper.kt
package no.nav.melosys.eessi.service.sed.mapper.fra_sed

import mu.KotlinLogging
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA003
import no.nav.melosys.eessi.service.sed.LandkodeMapper.mapTilNavLandkode

private val log = KotlinLogging.logger {}

abstract class FraSedA003Mapper : NyttLovvalgSedMapper<MedlemskapA003> {
    override fun hentLovvalgsland(medlemskap: MedlemskapA003): String? {
        return mapTilNavLandkode(medlemskap.vedtak!!.land)
    }

    override fun hentLovvalgsbestemmelse(medlemskap: MedlemskapA003): String? {
        return medlemskap.relevantartikkelfor8832004eller9872009
    }

    override fun sedErEndring(medlemskap: MedlemskapA003): Boolean {
        val erEndring = !"ja".equals(medlemskap.vedtak!!.eropprinneligvedtak, ignoreCase = true)
        log.info(
            "sedErEndring i A003 er $erEndring, med erendringsvedtak: ${
                medlemskap.vedtak!!.erendringsvedtak
            } og eropprinneligvedtak: ${medlemskap.vedtak!!.eropprinneligvedtak}",
        )
        return erEndring
    }

    override fun hentMedlemskap(sed: SED): MedlemskapA003 {
        return sed.medlemskap as MedlemskapA003
    }

    override fun erMidlertidigBestemmelse(medlemskap: MedlemskapA003): Boolean {
        return "ja" == medlemskap.isDeterminationProvisional
    }
}
